build-llvm.py

```diff
@@ -0,0 +1,402 @@
+#     defintions and then using printf to format them. In Python, that is a lot
+#     uglier because of how subprocess pipes work (see the 'linker_test'
+#     function below). While this isn't much better, it works and doesn't
+#     depend on pipes.
+#
+# How it works:
+#     Clang's version string is different on various distributions but they
+#     all similar enough.
+#
+#     apt.llvm.org: clang version 9.0.0-svn357849-1~exp1+0~20190406231252.1~1.gbpd89028 (trunk)
+#     Arch Linux: clang version 8.0.0 (tags/RELEASE_800/final)
+#     Debian: clang version 3.8.1-24 (tags/RELEASE_381/final)
+#     Fedora: clang version 7.0.1 (Fedora 7.0.1-6.fc29)
+#     Ubuntu: clang version 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
+#
+#     This one liner:
+#         1. Gets the first line of 'clang --version' (.splitlines()[0])
+#         2. Gets the third chunk, which contains the numeric verison (.split(" ")[2])
+#         3. Removes everything after a hyphen if it exists (.split("-")[0])
+#         4. Removes the periods (.replace(".", ""))
+#
+# FIXME: Is there a better way to do this?
+def clang_version(cc):
+    return int(
+        subprocess.check_output([
+            cc, "--version"
+        ]).decode("utf-8").splitlines()[0].split(" ")[2].split("-")[0].replace(
+            ".", ""))
+
+
+def parse_parameters(root):
+    parser = argparse.ArgumentParser(
+        formatter_class=argparse.RawTextHelpFormatter)
+    parser.add_argument("-b",
+                        "--branch",
+                        help=textwrap.dedent("""\
+                        By default, the script builds the master branch (tip of tree) of LLVM. If you would
+                        like to build an older branch, use this parameter. This may be helpful in tracking
+                        down an older bug to properly bisect. This value is just passed along to 'git checkout'
+                        so it can be a branch name, tag name, or hash.
+
+                        """),
+                        type=str,
+                        default="master")
+    parser.add_argument("-d",
+                        "--debug",
+                        help=textwrap.dedent("""\
+                        By default, the script builds LLVM in the release configuration with all of
+                        the tests turned off and optimization at O2. This disables that optimization,
+                        builds the tests, and changes the configuration to debug. This can help with
+                        reporting problems to LLVM developers but will make compilation of both LLVM
+                        and the kernel go slower.
+
+                        """),
+                        action="store_true")
+    parser.add_argument("-i",
+                        "--incremental",
+                        help=textwrap.dedent("""\
+                        By default, the script removes all build artifacts from previous compiles. This
```

Comment:
hmm...do we want the default to be incremental? I don't think so, but could be persuaded otherwise.

Reply:
I always run clean builds personally and many others I know do too. The only time I do an incremental build is if I am bisecting so I can get super quick compiles off.